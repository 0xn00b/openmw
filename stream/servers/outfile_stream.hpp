#ifndef MANGLE_OSTREAM_FILESERVER_H
#define MANGLE_OSTREAM_FILESERVER_H

#include "std_ostream.hpp"
#include <fstream>

namespace Mangle {
namespace Stream {

/** File stream based on std::ofstream, only supports writing.
 */
class OutFileStream : public StdOStream
{
  std::ofstream file;

 public:
  /**
     By default we overwrite the file. If append=true, then we will
     open an existing file and seek to the end instead.
   */
  OutFileStream(const std::string &name, bool append=false)
    : StdOStream(&file)
    {
      std::ios::openmode mode = std::ios::binary;
      if(append)
        mode |= std::ios::app;
      else
        mode |= std::ios::trunc;

      file.open(name.c_str(), mode);

      if(file.fail())
        throw str_exception("OutFileStream: failed to open file " + name);
    }
  ~OutFileStream() { file.close(); }
};

typedef boost::shared_ptr<OutFileStream> OutFileStreamPtr;

}} // namespaces
#endif
