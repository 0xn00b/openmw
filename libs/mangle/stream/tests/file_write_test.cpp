#include "../servers/outfile_stream.hpp"
#include <iostream>

using namespace Mangle::Stream;
using namespace std;

void print(Stream &str)
{
  cout << "size=" << str.size()
       << " pos=" << str.tell()
       << " eof=" << str.eof()
       << endl;
}

int main()
{
  {
    cout << "\nCreating file\n";
    OutFileStream out("test.file");
    print(out);
    out.write("hello",5);
    print(out);
  }

  {
    cout << "\nAppending to file\n";
    OutFileStream out("test.file", true);
    print(out);
    out.write(" again\n",7);
    print(out);
  }

  {
    cout << "\nOverwriting file\n";
    OutFileStream out("test.file");
    print(out);
    out.write("overwrite!\n",11);
    print(out);
  }
  return 0;
}
