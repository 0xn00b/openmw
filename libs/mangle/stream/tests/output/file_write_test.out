
Creating file
size=0 pos=0 eof=0
size=5 pos=5 eof=0

Appending to file
size=5 pos=5 eof=0
size=12 pos=12 eof=0

Overwriting file
size=0 pos=0 eof=0
size=11 pos=11 eof=0
